### PR TITLE
Test that scalar function with instanceFactory has bounded retained size

### LINKED ISSUE #8428: Test that functions using PageBuilder in instanceFactory use bounded memory
This is necessary because the PageBuilder API is quite error prone. And there had been multiple cases where bugs leads to query failures and GC pressures.

### PR DESCRIPTION
**Pull request by haozhun**
Fixes #8428

### CODE REVIEW

**Comment by wenleix** [comment]
Would be great if can introduce some motivation? Try to avoid humongous allocation because of the state class ?...

Also... every time I need to rethink what is "instance factory"...  IIRC it's "state class constructor" right? :D

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java [hunk 1]
```diff
@@ -268,6 +277,114 @@ private Object selectUniqueValue(String projection, Type expectedType, Session s
...
+                return sizeOf((short[]) object);
+            }
+            else if (type == Block[].class) {
+                Object[] objects = (Object[]) object;
+                return Arrays.stream(objects)
+                        .mapToLong(this::getRetainedSizeOf)
+                        .sum();
+            }
+            else {
+                throw new UnsupportedOperationException(format("Unknown type encountered: %s", type));
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java]: Throwing `IllegalArgumentException` seems like a better fit here, what do you think?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java [hunk 2]
```diff
@@ -268,6 +277,114 @@ private Object selectUniqueValue(String projection, Type expectedType, Session s
...
+            Field[] fields = type.getDeclaredFields();
+            for (Field field : fields) {
+                try {
+                    if (field.getType().isPrimitive()) {
+                        continue;
+                    }
+                    field.setAccessible(true);
+                    retainedSize += getRetainedSizeOf(field.get(object));
+                }
+                catch (IllegalAccessException | IllegalArgumentException t) {
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java]: Which call throws `IllegalArgumentException` (unless we make the above `UnsupportedOperationException` an `IllegalArgumentException`)?

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java [hunk 3]
```diff
@@ -268,6 +277,114 @@ private Object selectUniqueValue(String projection, Type expectedType, Session s
...
+        return retainedSize;
+    }
+
+    private long getRetainedSizeOf(Object object)
+    {
+        Class type = object.getClass();
+        if (type == PageBuilder.class) {
+            return ((PageBuilder) object).getRetainedSizeInBytes();
+        }
+        else if (type.isArray()) {
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java]: Don't need else here as previous if returns.

**Pull request merged**